File: leetcode/01_two_sum.ts
// Link: https://leetcode.com/problems/two-sum/

function twoSum(nums: number[], target: number): number[] {
    // create a map to store number -> index pairs
    const map = new Map()
    
    // iterate through the array once
    for (let i = 0; i < nums.length; i++) {
        // calculate the complement needed to reach target
        // [2, 7, 11, 15], target = 9
        // first pass: 9 - 2 = 7 -> store 2:7 in map
        // second pass: 9 - 7 = 2 -> 2 found in map, return [0, 1]
        const complement = target - nums[i]
        
        // check if complement exists in our map
        if (map.has(complement)) {
            // if found, return current index and complement's index
            return [map.get(complement), i]
        }
        
        // store current number and its index
        map.set(nums[i], i)
    }
    
    // return empty array if no solution found (per problem constraints, should never happen)
    return []
}

/* 
// alternate solution 

function twoSum(nums: number[], target: number): number[] {
    for(let i = 0; i < nums.length; i++) {
        for(let j = i + 1; j < nums.length; j++) {
            if(nums[i] + nums[j] === target) {
                return [i, j]
            }
        }
    }
    return []
};

*/

File: leetcode/01_two_sum.java
// Link: https://leetcode.com/problems/two-sum/

import java.util.HashMap;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        // create a hashmap to store number -> index pairs
        HashMap<Integer, Integer> map = new HashMap<>();
        
        // iterate through the array once
        for (int i = 0; i < nums.length; i++) {
            // calculate the complement needed to reach target
            int complement = target - nums[i];
            
            // check if complement exists in our map
            if (map.containsKey(complement)) {
                // if found, return current index and complement's index
                return new int[] {map.get(complement), i};
            }
            
            // store current number and its index
            map.put(nums[i], i);
        }
        
        // return empty array if no solution found (per problem constraints, should never happen)
        return new int[] {};
    }
}